src/NHibernate/Transform/DistinctEntitiesResultTransformer.cs

```diff
@@ -0,0 +1,128 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+using Iesi.Collections.Generic;
+
+namespace NHibernate.Transform
+{
+	[Serializable]
+	public class DistinctEntitiesResultTransformer : IResultTransformer
+	{
+		private static readonly IInternalLogger log = LoggerProvider.LoggerFor(typeof(DistinctEntitiesResultTransformer));
+		private static readonly object Hasher = new object();
+		private readonly bool[] returns;
+
+		internal sealed class Identity
+		{
+			internal readonly object entity;
+
+			internal Identity(object entity)
+			{
+				this.entity = entity;
+			}
+
+			public override bool Equals(object other)
+			{
+				Identity that = (Identity) other;
+				return ReferenceEquals(entity, that.entity);

```

Comment:
My proposed renaming would not suit this implementation, as it most likely to not handle properly equality of value types. Maybe the result transformer constructor should accept an `IEqualityComparer`.